### DISCUSSION TITLE
Things to port from mojo.js

### DISCUSSION THREAD

**Discussion by rabbiveesh** [mojolicious/mojo#1785] [Ideas]
@kraih asked on IRC that someone should keep a list of features to port for mojo.js

Here's the ones I found searching my IRC logs.

- new request id implementation
- log level `trace`
- exception formats

**Comment by kraih** [comment]
Router:
* Every route has its own weak ref to the root, removing the need to follow parent references
* New faster core router with much cheaper data structures for nested routes (not so easy to port, but probably a significant performance boost)

CLI:
* `get` and `routes` commands use colors

**Comment by kraih** [comment]
Calling it "txt exceptions" doesn't really describe it though. The context object (mojo.js separates context and controller) has an `exceptionFormat` property. So you can do `ctx.exceptionFormat = 'json'` to change the format of the built-in 404 and 500 handlers. The test client automatically changes it to `txt`.

Currently there is support for `html`, `txt` and `json` format, with different levels of details based on development mode being active. And only the `html` format will try to render custom templates provided by the developer. And there are multiple levels of helpers the user can replace with extensions (`notFound`, `exception`, `jsonNotFound`, `jsonException`...and so on).

**Comment by kraih** [comment]
Probably the most important performance fix was to have a prefix for static files (`/public`). That saves tons of stat syscalls. We could introduce this as an optional feature in Mojolicious.
